### PR TITLE
Initial scaffolding for the new linter, without any rules implemented yet

### PR DESCRIPTION
**Pull request by rictic**
Extracted by rictic from usergenic's https://github.com/Polymer/polymer-linter/pull/1 PR

### CODE REVIEW

File: src/linter.ts [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+    let warnings: Warning[] = [];
+    for (const file of files) {
+      let document: Document;
+      try {
+        document = await this.analyzer.analyze(file);
+        for (const rule of this.rules) {
+          warnings = warnings.concat(await rule.check(document));
+        }
+      } catch (error) {
+        if (error instanceof WarningCarryingException) {
```
Comment (usergenic) [inline src/linter.ts]: Can we catch per-rule?  That way all rules will try to run if one fails.

**Review by usergenic — CHANGES_REQUESTED** [review]

File: src/linter.ts (hunk 1 above)
Comment (usergenic) [reply src/linter.ts]: If not here, lets file issue and move that code in separate PR?

Comment (rictic) [reply src/linter.ts]: +1, reworked how we do error handling here, PTAL

File: src/linter.ts [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+   */
+  public async lint(files: string[]): Promise<Warning[]> {
+    let warnings: Warning[] = [];
+    const analysisResult = await this._analyzeAll(files);
+    const documents = analysisResult.documents;
+    const analysisWarnings = analysisResult.warnings;
+    warnings = warnings.concat(analysisWarnings);
+    for (const document of documents) {
+      for (const rule of this._rules) {
+        warnings = warnings.concat(await rule.check(document));
```
Comment (usergenic) [inline src/linter.ts]: Can we add a TODO (or implement?) here to catch exceptions for the `rule.check()` call?  I'd be nice if we had a standard warning to push like we do in the `_analyzeAll` call.

Comment (rictic) [reply src/linter.ts]: +1, done

**Review by usergenic — APPROVED** [review]
This is all sorts of LGTM

**Comment by rictic** [comment]
Thanks for the review!

**Pull request merged**